src/Features/VisualBasic/Portable/ConvertLinq/VisualBasicConvertLinqQueryToLinqMethodProvider.vb

```diff
@@ -0,0 +1,234 @@
+                Select Case (queryClause.Kind())
+                    Case SyntaxKind.FromClause
+                        Return ProcessFromClause(expression, DirectCast(queryClause, FromClauseSyntax))
+                    Case SyntaxKind.DistinctClause
+                        Return CreateInvocationExpression(expression, NameOf(Enumerable.Distinct))
+                    Case SyntaxKind.WhereClause
+                        Return CreateInvocationExpression(expression, NameOf(Enumerable.Where), DirectCast(queryClause, WhereClauseSyntax).Condition)
+                    Case SyntaxKind.SkipClause
+                        Return CreateInvocationExpression(expression, NameOf(Enumerable.Skip), DirectCast(queryClause, PartitionClauseSyntax).Count)
+                    Case SyntaxKind.TakeClause
+                        Return CreateInvocationExpression(expression, NameOf(Enumerable.Take), DirectCast(queryClause, PartitionClauseSyntax).Count)
+                    Case SyntaxKind.OrderByClause
+                        Return ProcessOrderByClause(expression, DirectCast(queryClause, OrderByClauseSyntax))
+                    Case SyntaxKind.SelectClause
+                        Return ProcessSelectClause(expression, DirectCast(queryClause, SelectClauseSyntax))
+                    Case SyntaxKind.GroupJoinClause
+                        Return Nothing ' Too complicated for query method syntax.
+                    Case SyntaxKind.SimpleJoinClause
+                        Return Nothing ' Too complicated for query method syntax.
+                    Case SyntaxKind.LetClause
+                        Return Nothing ' Too complicated for query method syntax.
+                    Case SyntaxKind.AggregateClause
+                        Return Nothing ' Too complicated for query method syntax.
+                    Case SyntaxKind.GroupByClause
+                        Return Nothing ' Group by in query and in method are not 100% equivalent. Consdier support of some group by in the future.
+                    Case Else
+                        Return Nothing
+                End Select
+            End Function
+
+            Private Function ProcessFromClause(parentExpression As ExpressionSyntax, fromClause As FromClauseSyntax) As ExpressionSyntax
+                Dim expression = parentExpression
+                Dim variables = fromClause.Variables
+                If variables.Count > 1 Then
+                    Return Nothing ' Do not support more than 1 variable. It complicates query methods.
+                End If
+
+                Dim variable = variables.First()
+                If parentExpression Is Nothing Then
+                    expression = variable.Expression
+
+                    If variable.AsClause IsNot Nothing Then
+                        Dim identifier = SyntaxFactory.IdentifierName(NameOf(Enumerable.Cast))
```

Comment:
>Dim identifier = SyntaxFactory.IdentifierName(NameOf(Enumerable.Cast)) [](start = 24, length = 70)

This looks wrong, I don't believe VB ever uses Cast to convert to the type stated in the As clause. It is used only as an attempt to convert collection to queryable type and only with object as type argument. Several methods can be implicitly called on the collection, sometimes in a sequence, look at CollectionRangeVariableSymbolInfo structure. All information from that structure is sufficient and required to properly do the translation. No need to look at the syntax to check whether it has As clause or not.